### PR TITLE
Show custom fields option only if post type supports it

### LINKED ISSUE #16333: Custom fields in custom post type not appearing
hi,
i was looking for a similar issue, but i found nothing. if there was already i apologize in advance.

**Describe the bug**
after creating a custom post type, i am not able to see the custom fields.

**To reproduce**
1. fresh install
2.
```
function register_custom_post_types() {
	register_post_type('test', [
		'labels' => [
			'name' => __('test'),
		],
		'public' => true,
		'supports' => [
			'title',
			'editor',
		],
		'show_in_rest' => true,
	]);
}
add_action('init', 'register_custom_post_types');
```
3. create a new cpt post and try to activate the custom fields in the options modal
4. (forced) refresh
5. nothing appears and the custom fields checkbox still unchecked

**Expected behavior**
the custom fields appear a the bottom of the editing page of the custom post type.

**Screenshots**
image

**Desktop:**
 - OS: windows
 - Browser firefox
 - Version 67.0.4 and 68.0b13

**Additional context**
wordpress 5.2.2

### PR DESCRIPTION
**Pull request by Soean**
## Description
Fixes #16333 
The custom field option should only be visible, if the post type supports this feature. The old solution doesn't work anymore, because `enableCustomFields ` has now a default value `false` (added in #14082).
So this new solution checks the supports array of the post type for the value `custom-fields`.

## How has this been tested?
Register a post type without custom field support, so the option should be hidden. If you add the custom fields support, the option should be visible.
```php
function register_custom_post_types() {
	register_post_type('test', [
		'labels' => [
			'name' => __('test'),
		],
		'public' => true,
		'supports' => [
			'title',
			'editor',
			// 'custom-fields',
		],
		'show_in_rest' => true,
	]);
}
add_action('init', 'register_custom_post_types');

```

## Screenshot
Bildschirmfoto 2019-06-27 um 21 48 31

### CODE REVIEW

File: packages/edit-post/src/components/options-modal/meta-boxes-section.js [hunk 1]
```diff
@@ -34,12 +34,13 @@ export function MetaBoxesSection( { areCustomFieldsRegistered, metaBoxes, ...sec
...
 export default withSelect( ( select ) => {
-	const { getEditorSettings } = select( 'core/editor' );
+	const { getPostType } = select( 'core' );
+	const { getEditedPostAttribute } = select( 'core/editor' );
 	const { getAllMetaBoxes } = select( 'core/edit-post' );
+	const postType = getPostType( getEditedPostAttribute( 'type' ) );
 
 	return {
-		// This setting should not live in the block editor's store.
-		areCustomFieldsRegistered: getEditorSettings().enableCustomFields !== undefined,
```
Comment (youknowriad) [inline packages/edit-post/src/components/options-modal/meta-boxes-section.js]: What's the purpose of the previous `enableCustomFields` setting?
I assume it's meant to be set in WordPress Core. is there code we need to remove from there?

Comment (Soean) [reply packages/edit-post/src/components/options-modal/meta-boxes-section.js]: `supportsCustomFields` defines if the post type supports this feature. `enableCustomFields` is the setting if this meta-box is visible. Before `enableCustomFields` had the default value `false` we checked with `!== undefined` if the feature exits. Now `enableCustomFields` can't be `undefined` anymore, so we need to check `supportsCustomFields`.

We still need `enableCustomFields` for checking if the meta-box should be visible. This logic is already merged into core.

Comment (youknowriad) [reply packages/edit-post/src/components/options-modal/meta-boxes-section.js]: Ok, makes sense. Changes look good to me. I'll wait for @talldan or @noisysocks to confirm.

**Review by noisysocks — CHANGES_REQUESTED** [review]
Thanks for looking into this!

Unfortunately we can't rely on checking if the CPT supports custom fields because plugins will disable custom fields by modifying the `$wp_meta_boxes` global during the `do_meta_boxes` action. 

You can see this problem by following these steps:

1. Check out this branch
2. Install the Advanced Custom Fields plugin
3. Create a new Field Group with a filed in it by going to _Custom Fields > Add New_
3. Create a new post by going to _Posts > Add New_
4. Select _More tools & options > Options_

Expected result: There should be no _Custom Fields_ checkbox because ACF replaces the Core Custom Fields functionality. 
Actual result: There's a _Custom Fields_ checkbox which doesn't do anything. 

This is why WordPress will check the `$wp_meta_boxes` global and unset the `enableCustomFields` editor setting (making it `undefined`) if custom fields were disabled by a plugin:

https://github.com/WordPress/wordpress-develop/blob/5fea17ba03c42d16a6517b164d3437bd53ac8ac9/src/wp-admin/edit-form-blocks.php#L366-L370

Are we able to stick with the `undefined|true|false` approach that existed before https://github.com/WordPress/gutenberg/pull/14082? I think that the error is in having the default setting for `enableCustomFields` be `false`. The default should be `undefined`.

**Comment by kosso** [comment]
Still having this issue on Version 5.3.2. 

My custom post type `supports` `array( 'title', 'editor', 'thumbnail', 'custom-fields' )`

And I can see the Options checkbox to activate 'Custom Fields'. 
But after clicking the provided button to view them in the panel, no change occurs after the page reloads. 

Am I missing something?

**Comment by paaljoachim** [comment]
Hello everyone.

What is needed to move this PR forward? 
Or perhaps a new PR is needed?

It would be great with an update @Soean 
Thank you!

@ntsekouras Nik perhaps this is a PR you are able to move forward?

**Comment by ntsekouras** [comment]
>@ntsekouras Nik perhaps this is a PR you are able to move forward?

Hey Paal. I think some other person with previous work on custom fields could be more effective here. I'll keep this in my queue though.

**Comment by skorasaurus** [comment]
I believe this can be closed because of the work done at https://github.com/WordPress/gutenberg/pull/33931; the custom fields preference does not display, when i create a custom post type ,following the directions in https://github.com/WordPress/gutenberg/issues/16333 in a new gutenberg build that includes #33931

**Pull request closed without merging**